### PR TITLE
Upp 1625 Image resolver

### PR DESCRIPTION
**Pull request by adelah**
New service which will return unrolled images for a given article UUID, by calling content-public-read /content endpoint

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+			Timeout: 10 * time.Second,
+			Transport: &http.Transport{
+				MaxIdleConnsPerHost: 100,
+				Dial: (&net.Dialer{
+					KeepAlive: 30 * time.Second,
+				}).Dial,
+			},
+		}
+		sc := content.ServiceConfig{
+			*port,
```
Comment (tamas-molnar) [inline main.go]: Please use named field initialization everywhere.

File: content/body_parser.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+	GetEmbedded(content Content) ([]string, error)
+}
+
+type BodyParser struct{}
+
+func (bp BodyParser) GetEmbedded(content Content) ([]string, error) {
+	var ids []string
+
+	if content.BodyXML == "" {
+		return ids, errors.New(fmt.Sprintf("Cannot parse empty body of content [%s]", content.UUID))
```
Comment (tamas-molnar) [inline content/body_parser.go]: You can use `fmt.Errorf(...)` here.

File: README.md [hunk 3]
```diff
@@ -1 +1,293 @@
 # image-resolver
+
+Image resolver is an internally used API for retrieving unrolled images.
+
```
Comment (peterschubert) [inline README.md]: missing integration badges

File: .gitignore [hunk 4]
```diff
@@ -0,0 +1,5 @@
+.idea/
+*.exe
+vendor/*/
+.vscode
+*.iml
```
Comment (tosan88) [inline .gitignore]: Maybe worth defining how a .gitignore standard file would look like in github.com/Financial-Times/cookiecutter-upp-golang

Comment (bogdanguranda) [inline .gitignore]: I definitely agree with having that in the template. I can do a PR for that and see what others think.

File: main.go [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+	})
+	graphiteTCPAddress := app.String(cli.StringOpt{
+		Name:   "graphite-tcp-address",
+		Value:  "",
+		Desc:   "Graphite TCP address, e.g. graphite.ft.com:2003. Leave as default if you do NOT want to output to graphite (e.g. if running locally)",
+		EnvVar: "GRAPHITE_TCP_ADDRESS",
+	})
+	graphitePrefix := app.String(cli.StringOpt{
+		Name:   "graphite-prefix",
+		Value:  "coco.services.$ENV.content-preview.0",
```
Comment (danielcopaciu) [inline main.go]: Image-resolver should have its own folder. I don't see the reason to default to content-preview.

File: main.go [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+	})
+	routerAddress := app.String(cli.StringOpt{
+		Name:   "routerAddress",
+		Value:  "localhost:8080",
+		Desc:   "Vulcan host",
+		EnvVar: "ROUTER_ADDRESS",
+	})
+	graphiteTCPAddress := app.String(cli.StringOpt{
+		Name:   "graphite-tcp-address",
+		Value:  "",
```
Comment (danielcopaciu) [inline main.go]: You can leave this undefined.

File: main_test.go [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+}
+
+func TestShouldReturn400InvalidJson(t *testing.T) {
+	startContentAPIMock("happy")
+	startImageResolverService()
+	defer stopServices()
+	var jsonStr = []byte(`{
+	"id": "http://www.ft.com/thing/22c0d426-1466-11e7-b0c1-37e417ee6c76",
+		"type": "http://www.ft.com/ontology/content/Article",
+		"blabla"}`)
```
Comment (danielcopaciu) [inline main_test.go]: Can this have a more significant value?

File: README.md [hunk 8]
```diff
@@ -1 +1,451 @@
...
+Usage: image-resolver [OPTIONS]
+
+Options:
+  --port="8080"                               Port to listen on ($PORT)
+  --cprHost="content-public-read"             The host to connect to content-public-read API
+  --routerAddress="localhost:8080"            Vulcan host
+  --graphiteTCPAddress=""                     Graphite TCP address, e.g. graphite.ft.com:2003. Leave as default if you d
+o NOT want to output to graphite (e.g. if running locally) ($GRAPHITE_ADDRESS)
+  --graphitePrefix=""                         Prefix to use. Should start with content, include the environment, and the
+ host name. e.g. coco.pre-prod.public-things-api.1 ($GRAPHITE_PREFIX)
```
Comment (danielcopaciu) [inline README.md]: We should the actual path of the service.

File: README.md [hunk 9]
```diff
@@ -1 +1,451 @@
...
+  "lastModified": "2017-03-31T15:42:35.266Z",
+  "canBeSyndicated": "yes",
+  "accessLevel": "subscribed",
+  "canBeDistributed": "yes"
+}
+```
+
+
+## Example 2 (alternative images)
+POST: /image-resolver/content
```
Comment (danielcopaciu) [inline README.md]: I think this should be /content only, without the name of the service.

File: content/model.go [hunk 10]
```diff
@@ -0,0 +1,12 @@
+package content
+
+type Content map[string]interface{}
+
+type PromotionalImage struct {
```
Comment (danielcopaciu) [inline content/model.go]: I don't think this struct is necessary.

File: content/model.go [hunk 11]
```diff
@@ -0,0 +1,12 @@
...
+	PromotionalImage interface{} `json:"promotionalImage,omitempty"`
+}
+
+type ImageOutput struct {
```
Comment (danielcopaciu) [inline content/model.go]: I don't think this struct is necessary. Please use Content instead, and save the type.

File: content/body_parser.go [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+	return &BodyParser{}
+}
+
+func (bp *BodyParser) GetEmbedded(body string) ([]string, error) {
+	var ids []string
+	ids, err := parseXMLBody(body)
+	return ids, err
+}
+
+func parseXMLBody(body string) ([]string, error) {
```
Comment (danielcopaciu) [inline content/body_parser.go]: Can you merge these two functions?

File: content/reader.go [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+		return result, fmt.Errorf("Error connecting to content-public-read for uuid=%s, err=%v", uuid, err)
+	}
+	defer res.Body.Close()
+
+	body, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return result, fmt.Errorf("Error reading response from content-public-read for uuid=%s, err=%v", uuid, err)
+	}
+
+	err = json.Unmarshal(body, &result)
```
Comment (danielcopaciu) [inline content/reader.go]: Why don't you check the status code first? If it's not equal to 200 you can skip the unmarshalling.

File: content/service.go [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+		result[LeadImages] = ir.getLeadImages(images)
+	}
+	return result
+}
+
+func (ir *ImageResolver) getImage(uuid string) []Content {
+	return ir.getImageSets([]string{uuid})
+}
+
+func (ir *ImageResolver) getEmbeddedImages(UUIDs []string) []Content {
```
Comment (danielcopaciu) [inline content/service.go]: Can you remove this function?

File: content/service.go [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+func (ir *ImageResolver) getImageSets(uuids []string) []Content {
+	outputs := []Content{}
+	for _, uuid := range uuids {
+		id := extractIdfromUrl(uuid)
+		imageSet, err := ir.reader.Get(id)
+		if err != nil {
+			imageSet = make(map[string]interface{})
+			imageSet[ID] = uuid
+			outputs = append(outputs, imageSet)
+			log.Infof("Error unrolling image uuid = %s, err: %v", id, err)
```
Comment (danielcopaciu) [inline content/service.go]: I think the level should be error.

File: content/service.go [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+	result := make([]Content, len(leadImages))
+	for index, leadImg := range leadImages {
+		result[index] = make(map[string]interface{})
+		img := leadImg.(map[string]interface{})
+		id := img[ID].(string)
+		im, err := ir.reader.Get(extractIdfromUrl(id))
+		if err != nil{
+			result[index][ID]= id
+			result[index][Type] = img[Type].(string)
+			log.Infof("Error unrolling leadimage uuid =%s, err=%v", id, err)
```
Comment (danielcopaciu) [inline content/service.go]: The level should be error.

File: content/body_parser_test.go [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+func TestBodyEmpty(t *testing.T) {
+	var reader Reader
+	var parser Parser
+	var ir ImageResolver
+	reader = NewContentReader("", "")
+	parser =NewBodyParser()
+	ir = *NewImageResolver(&reader, &parser)
+	str := ""
```
Comment (danielcopaciu) [inline content/body_parser_test.go]: It's not necessary to initialize a variable with it's default value

File: content/handlers.go [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
+
+	var article Content
+	var id string
+	tid := transactionidutils.GetTransactionIDFromRequest(r)
+
+	hh.Log.TransactionStartedEvent(r.RequestURI, tid, id)
+
+	b, err := ioutil.ReadAll(r.Body)
+	if err != nil {
```
Comment (danielcopaciu) [inline content/handlers.go]: This seems to be duplicated. Can you create a separate function?

File: content/handlers.go [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+		hh.Log.RequestFailedEvent(AppName, r.RequestURI, tid, http.StatusBadRequest, id)
```
Comment (danielcopaciu) [inline content/handlers.go]: This seems to be duplicated. Can you create a separate function?

File: content/service.go [hunk 20]
```diff
@@ -0,0 +1,181 @@
...
+	bodyXml, found := body[BodyXml].(string)
+	if found && bodyXml != "" {
+		emImagesUUIDs, err := ir.parser.GetEmbedded(bodyXml)
+		if err == nil {
+			embeddedImages := ir.getEmbeddedImages(emImagesUUIDs)
+			if len(embeddedImages) > 0 {
+				result[Embeds] = embeddedImages
+			}
+		} else {
+			log.Infof("Error parsing body for article uuid=%s, err=%v", body[ID].(string), err)
```
Comment (danielcopaciu) [inline content/service.go]: The level should be error

Comment (danielcopaciu) [inline content/service.go]: The level should be error

File: content/service.go [hunk 21]
```diff
@@ -0,0 +1,181 @@
...
+}
+
+func (ir *ImageResolver) getImageSetMembers(membersUUIDs []string) []Content {
+	members := []Content{}
+	for _, member := range membersUUIDs {
+		id := extractIdfromUrl(member)
+		im, err := ir.reader.Get(id)
+		if err != nil {
+			im[ID] = member
+			log.Infof("Error unrolling image uuid =%s, err=%v", member, err)
```
Comment (danielcopaciu) [inline content/service.go]: The level  should be error

File: content/service.go [hunk 22]
```diff
@@ -0,0 +1,181 @@
...
+			result[index][Image] = im
+			result[index][Type] = img[Type].(string)
+			result[index][ID] = id
+		}
+	}
+	return result
+}
+
+func extractIdfromUrl(url string) string {
+	id := ""
```
Comment (danielcopaciu) [inline content/service.go]: there is no need to initialize a variable with it's default value

Comment (danielcopaciu) [inline content/service.go]: There is no need to initialize a variable with its default value

**Pull request merged**